### PR TITLE
OCPBUGS-64825: Add MachineConfigNode informer to trigger MCP machine count syncs

### PR DESCRIPTION
**Pull request by isabella-janssen**
Closes: OCPBUGS-64825

**- What I did**
This adds a MachineConfigNode informer in the node controller so that changes to a node's MCN trigger syncs of the MachineConfigPool machine counts. This change is necessary with the introduction of #5141, which calculates the updated and degraded machine counts with MCN conditions instead of node annotations. It also adds a `TestCalculateStatusWithImageModeReporting` test, which tests the same cases as the previously existing `TestCalculateStatus` test, but in a tech preview situation.

**- How to verify it**
This can be tested locally by launching a 4.21 tech preview cluster with this PR included. It can also be tested by ensuring that all existing `MachineConfigNode` and `ImageModeStatusReporting` pass--specifically the `Should properly report MCN conditions on node degrade` test mentioned in the attached bug.

**- Description for the changelog**
OCPBUGS-64825: Add MachineConfigNode informer to trigger MCP machine count syncs

### CODE REVIEW

**Comment by isabella-janssen** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-1of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-2of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-3of3

**Comment by isabella-janssen** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-1of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-2of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-3of3

**Comment by isabella-janssen** [comment]
/payload-job periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-1of2 periodic-ci-openshift-machine-config-operator-release-4.21-periodics-e2e-aws-mco-disruptive-techpreview-2of2 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-1of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-2of3 periodic-ci-openshift-release-master-ci-4.21-e2e-aws-ovn-techpreview-serial-3of3

**Comment by isabella-janssen** [comment]
> Generally looks good, would adding unit test(s) for this case be viable?

@djoshy I was considering adding unit tests but did not for two reasons:

1. The `Should properly report MCN conditions on node degrade` regression test makes sure the MCP reacts to degrade corrections and the `MachineConfigPool machine counts should transition correctly on an update in a default MCP` and `MachineConfigPool machine counts should transition when OCB is enabled in a default MCP` regression tests make sure the MCP reacts during standard updates.
2. The informers "reacting" only reflects if it's working in tech preview and I'm not sure we can distinguish test cases based on tech preview feature gates?

I think unit tests would be a good idea because signal from (1) takes a lot longer than the feedback from unit tests, but I'd need to better understand how to handle feature gate "enablement" in unit tests to remediate (2). Do you have an example that would help solve (2)?

**Comment by djoshy** [comment]
> The informers "reacting" only reflects if it's working in tech preview and I'm not sure we can distinguish test cases based on tech preview feature gates?

We have a few unit tests that initializes controllers with [feature gates enabled](https://github.com/openshift/machine-config-operator/blob/main/pkg/controller/node/status_test.go#L772-L776). I don't think we need to necessarily try to test the informer reaction since that's pretty standard(although that would be great if we want to add a general unit test for the syncs), I was thinking something more surgical in `TestCalculateStatus` or so to cover MCN edge cases such as these(plus the regular "happy" path cases). These checks/counts are becoming quite nuanced so I worry about not having an immediate signal until our CR goes red.

File: pkg/controller/node/status_test.go [hunk 1]
```diff
@@ -770,14 +833,14 @@ func TestCalculateStatus(t *testing.T) {
 				},
 			}
 			f := newFixtureWithFeatureGates(t,
-				[]apicfgv1.FeatureGateName{
+				[]configv1.FeatureGateName{
 					features.FeatureGateMachineConfigNodes,
 					features.FeatureGatePinnedImages,
```
Comment (djoshy) [inline pkg/controller/node/status_test.go:1322]: Shouldn't we pass `t.featureGates` through here? Where is it being used in the controller?

Comment (isabella-janssen) [reply pkg/controller/node/status_test.go:1322]: I have a commit fixing this oversight I had not push up yet since I had some other failures I was working to correct. I'll let you know when the UT updates are ready.

Comment (djoshy) [reply pkg/controller/node/status_test.go:1322]: ack, no worries!

Comment (isabella-janssen) [reply pkg/controller/node/status_test.go:1322]: The units should be all ready now, @djoshy! I made them a separate test since the setup between the TP and non-TP cases were different enough that it got a bit confusing when they were generalized into one.

**Comment by djoshy** [comment]
The units look great, and thanks for naming them and reorganizing them!

/lgtm

**Comment by sergiordlr** [comment]
We can see the test cases passing. Especially the one previously failing and mentioned in the bug:
` [sig-mco][OCPFeatureGate:MachineConfigNodes] [Suite:openshift/machine-config-operator/disruptive][Disruptive]Should properly report MCN conditions on node degrade [apigroup:machineconfiguration.openshift.io] [Serial]	4m47s
`

/label qe-approved
/verified by  payload execution

**Comment by isabella-janssen** [comment]
/label acknowledge-critical-fixes-only

This fix should be safe to merge (the required payloads this impacts passed in the [rehearsal run](https://github.com/openshift/machine-config-operator/pull/5422#issuecomment-3554004119)) and is needed to get the 4.21 MCO disruptive suite performance a bit more healthy, which we need to have stable to promote the suite's job tier to report to component readiness.
